dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,658 @@
+    meta.update(score=score)
+    assert meta["iterations"] > 0
+    meta["mean_score_time"] += time() - score_start
+    meta["mean_test_score"] = score
+    meta["time_score_evaluated"] = time() - start
+    return meta
+
+
+def _to_promote(result, completed_jobs, eta=None, asynchronous=True):
+    bracket_models = [
+        r
+        for r in completed_jobs
+        if r["bracket_iter"] == result["bracket_iter"] and
+        r["bracket"] == result["bracket"]
+    ]
+    to_keep = len(bracket_models) // eta
+
+    bracket_completed = len(bracket_models) == result["num_models"]
+    if not asynchronous and not bracket_completed:
+        to_keep = 0
+    if to_keep <= 1:
+        return []
+
+    def _promote_job(job):
+        job["num_models"] = to_keep
+        job["partial_fit_calls"] *= eta
+        job["bracket_iter"] += 1
+        return job
+
+    top = toolz.topk(to_keep, bracket_models, key="score")
+    if not asynchronous:
+        for job in top:
+            job = _promote_job(job)
+        return top
+    else:
+        if to_keep == 0:
+            return [result]
+        if result in top:
+            result = _promote_job(result)
+            return [result]
+        return []
+
+
+def _create_model(model, params, random_state=42):
+    model = clone(model).set_params(**params)
+    if "random_state" in model.get_params():
+        model.set_params(random_state=random_state)
+    return model, None  # right now no meta information
+
+
+def _model_id(s, n_i):
+    return "bracket={s}-{n_i}".format(s=s, n_i=n_i)
+
+
+@gen.coroutine
+def _hyperband(
+    model,
+    params,
+    X,
+    y,
+    max_iter=None,
+    eta=None,
+    fit_params={},
+    random_state=42,
+    test_size=None,
+    scorer=None,
+    asynchronous=None,
+):
+    client = default_client()
+    rng = check_random_state(random_state)
+    N, R, brackets = _get_hyperband_params(max_iter, eta=eta)
+    #  params = iter(ParameterSampler(params, n_iter=sum(N), random_state=rng))
+    params = [
+        ParameterSampler(params, 1, random_state=rng.randint(sum(N)))
+        for _ in range(sum(N))
+    ]
+    params = iter([list(p)[0] for p in params])
+
+    params = {
+        _model_id(s, n_i): next(params)
+        for s, n, r in zip(brackets, N, R)
+        for n_i in range(n)
+    }
+    model_futures = {
+        _model_id(s, n_i): client.submit(
+            _create_model,
+            model,
+            params[_model_id(s, n_i)],
+            random_state=rng.randint(100 * sum(N)),
+        )
+        for s, n, r in zip(brackets, N, R)
+        for n_i in range(n)
+    }
+
+    # lets assume everything in fit_params is small and make it concrete
+    fit_params = yield client.compute(fit_params)
+
+    r = train_test_split(X, y, test_size=test_size, random_state=rng)
+    X_train, X_test, y_train, y_test = r
+    if isinstance(X, da.Array):
+        X_train = futures_of(X_train.persist())
+        X_test = client.compute(X_test)
```

Comment:
So it's assumed that `X_test` fits in memory on the client machine? Is this required? If so, is that documented?

Reply:
Or perhaps this is a subset of the data...

Reply:
That is assumed, and documented. It is required (at least in this implementation) because this will be passed to `_score`.